src/core/src/signature.rs

```diff
@@ -20,138 +20,279 @@ use crate::sketch::Sketch;
+            if !self.is_protein {
+                // Setting the parameters only in the first iteration
+                if !self._dna_configured {
+                    self._dna_ksize = self.k_size as usize;
+                    self._dna_len = self.sequence.len();
+                    if self._dna_len < self._dna_ksize {
+                        return None;
                     }
-                    last_position_check += 1;
+                    // pre-calculate the reverse complement for the full sequence...
+                    self._dna_rc = revcomp(&self.sequence);
+                    self._dna_configured = true;
                 }
-                true
-            };
 
-            for i in 0..=len - ksize {
-                // ... and then while moving the k-mer window forward for the sequence
-                // we move another window backwards for the RC.
-                //   For a ksize = 3, and a sequence AGTCGT (len = 6):
-                //                   +-+---------+---------------+-------+
-                //   seq      RC     |i|i + ksize|len - ksize - i|len - i|
-                //  AGTCGT   ACGACT  +-+---------+---------------+-------+
-                //  +->         +->  |0|    2    |       3       |   6   |
-                //   +->       +->   |1|    3    |       2       |   5   |
-                //    +->     +->    |2|    4    |       1       |   4   |
-                //     +->   +->     |3|    5    |       0       |   3   |
-                //                   +-+---------+---------------+-------+
-                // (leaving this table here because I had to draw to
-                //  get the indices correctly)
-
-                let kmer = &sequence[i..i + ksize];
-
-                if !is_valid_kmer(i) {
-                    if !force {
-                        // throw error if DNA is not valid
-                        return Err(Error::InvalidDNA {
-                            message: String::from_utf8(kmer.to_vec()).unwrap(),
-                        });
+                // Processing DNA
+                if self.hash_function.dna() {
+                    let kmer = &self.sequence[self.kmer_index..self.kmer_index + self._dna_ksize];
+
+                    for j in std::cmp::max(self.kmer_index, self._dna_last_position_check)
+                        ..self.kmer_index + self._dna_ksize
+                    {
+                        if !VALID[self.sequence[j] as usize] {
+                            if !self.force {
+                                return Some(Err(Error::InvalidDNA {
+                                    message: String::from_utf8(kmer.to_vec()).unwrap(),
+                                }));
+                            } else {
+                                self.kmer_index += 1;
+                                // Move the iterator to the next step
+                                // This is a recursion, will cause a stack overflow in long seqs with non-ACGT
```

Comment:
Is this a recursion?

Reply:
It was, just missed removing the comment.

https://github.com/sourmash-bio/sourmash/pull/1653/commits/8367bee8e0537fa39abad40dc448f4344eb82497#diff-92fdf93a9e262cab04fd462dcc4d40b11ea94e72fe5911f398e00b0dd6b75292L123

Reply:
Resolved in [`9bed5d9` (#1653)](https://github.com/sourmash-bio/sourmash/pull/1653/commits/9bed5d9cf448639647c159562cb585b12b505f6b)